src/adapter/src/coord/appends.rs

```diff
@@ -892,4 +942,171 @@ impl GroupCommitWaiter {
 /// Note: We sometimes want to throttle how many group commits are running at once, which this
 /// permit allows us to do.
 #[derive(Debug)]
-pub struct GroupCommitPermit(#[allow(dead_code)] OwnedSemaphorePermit);
+pub struct GroupCommitPermit {
+    /// How long we want this permit to last before dropping.
+    hold_until: Instant,
+    /// Histogram used to measure how long we delayed for, if we delayed.
+    hold_histogram: Histogram,
+    /// Permit that is preventing other group commits from running.
+    ///
+    /// Only `None` if the permit has been moved into a tokio task for waiting.
+    _permit: Option<OwnedSemaphorePermit>,
+}
+
+impl Drop for GroupCommitPermit {
+    fn drop(&mut self) {
+        let now = Instant::now();
+
+        // We don't want to drop this permit yet so move it into a task that
+        // will wait until we reached our desired duration.
+        if self.hold_until > now {
+            let hold_until = self.hold_until.clone();
+            let permit = self._permit.take();
+
+            let hold_duration = hold_until.duration_since(now);
+            self.hold_histogram.observe(hold_duration.as_secs_f64());
+            tracing::debug!(?hold_duration, "defering drop of GroupCommit permit");
+
+            mz_ore::task::spawn(|| "group-commit-permit-waiter".to_string(), async move {
+                tokio::time::sleep_until(hold_until.into()).await;
+                // Explicitly drop our permit once our timeout has elapsed.
+                drop(permit);
+                tracing::debug!(?hold_duration, "drop GroupCommit permit");
+            });
+        }
+    }
+}
+
+/// When we start a [`Session`] we need to update some builtin tables, we don't want to wait for
+/// these writes to complete for two reasons:
+///
+/// 1. Doing a write can take a relatively long time.
+/// 2. Decoupling the write from the session start allows us to batch multiple writes together, if
+///    sessions are being created with a high frequency.
+///
+/// So as an optimization we do not wait for these writes to complete. But if a [`Session`] tries
+/// to query any of these builtin objects, we need to block that query on the writes completing to
+/// maintain linearizability.
+pub(crate) fn waiting_on_startup_appends(
+    catalog: &Catalog,
+    session: &mut Session,
+    plan: &Plan,
+) -> Option<(BTreeSet<CatalogItemId>, BoxFuture<'static, ()>)> {
+    let depends_on = match plan {
```

Comment:
I think we also need to call `transitive_uses` on these ids, in order to traverse referenced view / mat view definitions.

Reply:
I think we're okay for now since we don't support creating Views or MatViews on top of objects in the `mz_internal` schema, but I added a TODO about this and a soft assert to make sure all the tables we depend on are in `mz_internal`.

Happy to make the change anyways if you'd prefer, although initially it feels a bit bad looking up all of the transitive IDs when we know it won't matter?

Reply:
Ok, this is fine for now! Thanks for adding the assert!

Btw. there is a third case in which this could go wrong besides the cases mentioned in the added comment: We might at some point create a _builtin_ view on top of `mz_sessions` (which _is_ allowed), and a user might query that builtin view. But I've checked that we currently don't have any builtin that would depend on `mz_sessions`.

Reply:
Ahh good callout! I'll keep that in mind 👍